aten/src/ATen/native/cpu/ActivationKernel.cpp

```diff
@@ -0,0 +1,484 @@
+            tmpsum.store(tmp_sum_arr);
+            for (int64_t i = 0; i < Vec::size; i++) 
+              tmp_sum_scalar[j] += tmp_sum_arr[i];
+            //std::cout << "tmp_sum_scalar[" << j << "]: " << tmp_sum_scalar[j] << std::endl;
+          }
+          for (int64_t j = 0; j < inner_loop_end; j += Vec::size) {
+            int64_t leftover = _leftover<Vec::size>(j, inner_loop_end);
+            //std::cout << "j: " << j << " - " << leftover << std::endl;
+            Vec tmp_sum_scalar_vec;
+            tmp_sum_scalar_vec.load(tmp_sum_scalar + j, leftover);
+            //std::cout << "tmp_sum_scalar_vec1: " << tmp_sum_scalar_vec << std::endl;
+            tmp_sum_scalar_vec = tmp_sum_scalar_vec.log();
+            //std::cout << "tmp_sum_scalar_vec2: " << tmp_sum_scalar_vec << std::endl;
+            Vec max_input_vec;
+            max_input_vec.load(max_input_arr + j, leftover);
+            tmp_sum_scalar_vec = max_input_vec + tmp_sum_scalar_vec;
+            tmp_sum_scalar_vec.store(tmp_sum_scalar + j, leftover);
+          }
+          for (int64_t j = 0; j < inner_loop_end; j++) {
+            int64_t i = ii + j;
+            scalar_t* input_data = input_data_base + i * dim_size;
+            scalar_t* output_data = output_data_base + i * dim_size;
+            Vec sub_val_vec(tmp_sum_scalar[j]);
+            for (int64_t d = 0; d < dim_size; d += Vec::size) {
+              Vec value;
+              int64_t leftover = _leftover<Vec::size>(d, dim_size);
+              value.load(input_data + d, leftover);
+              value = value - sub_val_vec;
+              value.store(output_data + d, leftover);
+            }
+          }
+        }
+      },
+      ap);
+}
+
+template <typename scalar_t>
+inline void _vec_softmax_lastdim(
+    scalar_t* input_data_base,
+    scalar_t* output_data_base,
+    int64_t outer_size,
+    int64_t dim_size) {
+  int64_t grain_size = internal::TBB_GRAIN_SIZE / dim_size;
+  if (grain_size < 1)
+    grain_size = 1;
+
+  using Vec = vec256::Vec256<scalar_t>;
+  tbb::parallel_for(
+      tbb::blocked_range<int64_t>(0, outer_size, grain_size),
+      [&](const tbb::blocked_range<int64_t>& r) {
+        for (int64_t i = r.begin(); i < r.end(); i++) {
+          scalar_t* input_data = input_data_base + i * dim_size;
+          scalar_t* output_data = output_data_base + i * dim_size;
+
+          scalar_t max_input = _vec_max(dim_size, input_data);
+          scalar_t tmp_sum = _vec_norm_exp_sum_write(
+              output_data, dim_size, input_data, max_input);
+          tmp_sum = 1 / tmp_sum;
+          Vec sub_val_vec(tmp_sum);
+          for (int64_t d = 0; d < dim_size; d += Vec::size) {
+            Vec value;
+            int64_t leftover = _leftover<Vec::size>(d, dim_size);
+            value.load(output_data + d, leftover);
+            value = value * sub_val_vec;
+            value.store(output_data + d, leftover);
+          }
+        }
+      },
+      ap);
```

Comment:
Lol why is this one sooooooo much shorter than LogSoftmax? Why can't we have a templatized kernel like in the `host_softmax` case?

Reply:
Because log_softmax needs to vectorize the std::log call.